src/core/Oscillator.cpp

```diff
@@ -88,6 +106,238 @@ void Oscillator::update( sampleFrame * _ab, const fpp_t _frames,
 	}
 }
 
+void Oscillator::generateSineWaveTable(sample_t * table)
+{
+	// sinewaves are the only continuous waves containing no harmonics
+	// hence by definition all wavetables contain a single signal
+	// https://en.wikipedia.org/wiki/Sine_wave
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] = sinf(((float)i / (float)OscillatorConstants::WAVETABLE_LENGTH) * F_2PI);
+	}
+}
+
+void Oscillator::generateSawWaveTable(int bands, sample_t * table)
+{
+	// sawtooth wave contain both even and odd harmonics
+	// hence sinewaves are added for all bands
+	// https://en.wikipedia.org/wiki/Sawtooth_wave
+	float max = 0;
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] = 0.0;
+		for (float n = 1; n <= bands; n++)
+		{
+			table[i] += ((int)n%2 ? 1.0f : -1.0f) / n * sinf(F_2PI * i * n / (float)OscillatorConstants::WAVETABLE_LENGTH);
+		}
+		max = fmax(max, table[i]);
+	}
+
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] /= max;
+	}
+}
+
+
+void Oscillator::generateTriangleWaveTable(int bands, sample_t * table)
+{
+	// triangle waves contain only odd harmonics
+	// hence sinewaves are added for alternate bands
+	// https://en.wikipedia.org/wiki/Triangle_wave
+	float max = 0;
+	for (int i = 0 ; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] = 0.0;
+		for (float n = 0; n <= bands * 0.5; n++)
+		{
+			table[i] += ((int)n%2 ? -1.0f : +1.0f) / powf((float)(2 * n + 1),
+							   2.0f) *
+					sinf(F_2PI * (2.0f * n + 1) * i / (float)OscillatorConstants::WAVETABLE_LENGTH);
+		}
+		max = fmax(max, table[i]);
+	}
+
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] /= max;
+	}
+}
+
+void Oscillator::generateSquareWaveTable(int bands, sample_t * table)
+{
+	// square waves only contain odd harmonics,
+	// at diffrent levels when compared to triangle waves
+	// https://en.wikipedia.org/wiki/Square_wave
+	float max = 0;
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] = 0.0f;
+		for (float n = 1; n <= bands; n += 2)
+		{
+			table[i] += (1.0f / n) * sinf(F_2PI * i * n / OscillatorConstants::WAVETABLE_LENGTH);
+		}
+		max = fmax(max, table[i]);
+	}
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH; i++)
+	{
+		table[i] /= max;
+	}
+}
+
+
+
+//expects sample in sample buffer
+void Oscillator::generateFromFFT(int bands, float threshold, sample_t * table)
+{
+	//set unrequired bands to zero
+	for (int i = bands; i < OscillatorConstants::WAVETABLE_LENGTH * 2 + 1 - bands; i++)
+	{
+		s_specBuf[i][0] = 0.0f;
+		s_specBuf[i][1] = 0.0f;
+	}
+	for (int i = 0; i < OscillatorConstants::WAVETABLE_LENGTH * 2 + 1; ++i)
```

Comment:
I would guess the thresholds are pretty much arbitrary and in theory "no limit" would be best + slow, but I'll try to experiment to see how it changes the result. (I remember tweaking the values for some reason, but I don't remember why.)

Reply:
Found something in the original PR (second half of the post): https://github.com/LMMS/lmms/pull/4397#issuecomment-639083065

Still does not really explain why a threshold is needed at all, though.

Reply:
So, starting with G2, the exponential wave starts cutting off high harmonics. At C1, it essentially does not have any frequency higher than 5 kHz. I'm not sure how audible it would be, but it definitely looks weird in the spectrum (a sharp cutoff in an unexpected place).

I wonder if the cutoff could have some mixing benefit -- i.e., if a frequency is almost inaudible, do not even generate it → the frequency range will be left even cleaner for another instrument that will likely occupy the space. But such potential benefit seems not enough to produce weirdness like this in some cases:
![screen](https://user-images.githubusercontent.com/20932751/121092678-3c21f600-c7ec-11eb-97d1-f8025dc02514.png)

So I would consider removing the thresholds if it does not break anything..

(Also, nice: the GitHub medium-dark theme uses almost the same BG color as LMMS :) )

Reply:
If a user does want a cutoff, I rather think they should use LP filters. A threshold could be a "nice option" for a perfect lowpass, but I think it should not be forced (and not be the default).

No idea what causes the weirdness at 15K... Maybe we'll understand if we see the same spectrum without threshold?

Reply:
So, I set the limits to zero and this is how C3 exponential wave now looks:
![screen](https://user-images.githubusercontent.com/20932751/121958124-9a5a5600-cd63-11eb-8a25-60591b8c37bc.png)

I would say it is clear which version is better. The only question is whether to just set the threshold to zero, or remove all the code that checks is altogether, since it does not seems to be of any use.

Reply:
Yeah, it's clear that the version without threshold is much better.

I'd rather remove the code. Even if someone wanted that, as you have shown, the spectrum looks really wrong. And it should not be hard to re-add the threshold if it should ever be requested.

Reply:
Already removed in the previous commit, so marking as resolved.